File: primer_corte/Prototype/src/AbstractDocument.java
// Clase abstracta que implementa la interfaz Document
abstract class AbstractDocument implements Document {
    private String author;
    private String creationDate;
    private String title;
    
    // Constructor normal
    public AbstractDocument() {
    }
    
    // Constructor prototipo
    public AbstractDocument(AbstractDocument source) {
        if (source != null) {
            this.author = source.author;
            this.creationDate = source.creationDate;
            this.title = source.title;
        }
    }
    
    // Getters y setters
    public String getAuthor() {
        return author;
    }
    
    public void setAuthor(String author) {
        this.author = author;
    }
    
    public String getCreationDate() {
        return creationDate;
    }
    
    public void setCreationDate(String creationDate) {
        this.creationDate = creationDate;
    }
    
    public String getTitle() {
        return title;
    }
    
    public void setTitle(String title) {
        this.title = title;
    }
    
    @Override
    public abstract Document clone();
    
    @Override
    public abstract void print();
}

File: primer_corte/PrototypePython/abstract_document.py
from abc import ABC, abstractmethod
import copy
from typing import Dict

from document import Document

# Clase abstracta que implementa la interfaz Document
class AbstractDocument(Document):
    def __init__(self):
        self.author = None
        self.creation_date = None
        self.title = None

    # Constructor prototipo
    def __init_copy__(self, source=None):
        if source:
            self.author = source.author
            self.creation_date = source.creation_date
            self.title = source.title

    # Getters y setters
    def get_author(self):
        return self.author

    def set_author(self, author):
        self.author = author

    def get_creation_date(self):
        return self.creation_date

    def set_creation_date(self, creation_date):
        self.creation_date = creation_date

    def get_title(self):
        return self.title

    def set_title(self, title):
        self.title = title

    @abstractmethod
    def clone(self):
        pass

    @abstractmethod
    def print(self):
        pass